client/client.go

```diff
@@ -12,83 +15,98 @@ import (
 	"github.com/sajari/docconv"
 )
 
-var (
-	defaultParam    = "input"
-	defaultEndpoint = "http://localhost:8888/convert"
-)
+// DefaultEndpoint is the default endpoint for a docconv HTTP
+// server.
+const DefaultEndpoint = "localhost:8888"
+
+// DefaultHTTPClient is the default HTTP client used to make
+// all requests.
+var DefaultHTTPClient = http.DefaultClient
+
+// Opt is an option used in New to create Clients.
+type Opt func(*Client)
 
-type client struct {
-	endpoint string
-	param    string
-	client   http.Client
+// WithEndpoint set the endpoint on a Client.
+func WithEndpoint(endpoint string) Opt {
+	return func(c *Client) {
+		c.endpoint = endpoint
+	}
 }
 
-// Returns a client that will send to the default parameters used for the docconv pkg
-func NewDefaultClient() *client {
-	return &client{
-		endpoint: defaultEndpoint,
-		param:    defaultParam,
-		client:   http.Client{},
+// WithHTTPClient sets the *http.Client used for all underlying
+// calls.
+func WithHTTPClient(client *http.Client) Opt {
+	return func(c *Client) {
+		c.httpClient = client
 	}
 }
 
-// Returns a client using a custom endpoint to send to. The param indicates the
-// name of the parameter that the file will be send as in the form
-func NewClient(endpoint, param string) (*client, error) {
-	if endpoint == "" {
-		return nil, errors.New("You must specify an endpoint in the format domain:port")
+// New creates a new docconv client for interacting with a docconv HTTP
+// server.
+func New(opts ...Opt) *Client {
+	c := &Client{
+		endpoint:   DefaultEndpoint,
+		httpClient: DefaultHTTPClient,
 	}
-	if defaultParam == "" {
-		return nil, errors.New("You must specify an input param, the default is 'input'")
+
+	for _, opt := range opts {
+		opt(c)
 	}
-	return &client{
-		endpoint: defaultEndpoint,
-		param:    defaultParam,
-		client:   http.Client{},
-	}, nil
+	return c
 }
 
-// Convert a file from a local path using the http client
-func (c *client) Convert(path string) (*docconv.Response, error) {
+// Client is a docconv HTTP client.  Use New to make new Clients.
+type Client struct {
+	endpoint   string
+	httpClient *http.Client
+}
 
-	file, err := os.Open(path)
+// Convert a file from a local path using the http client
+func (c *Client) Convert(r io.Reader, filename string) (*docconv.Response, error) {
+	buf := &bytes.Buffer{}
+	w := multipart.NewWriter(buf)
+	part, err := w.CreateFormFile("input", filename)
 	if err != nil {
 		return nil, err
 	}
-
-	body := &bytes.Buffer{}
-	writer := multipart.NewWriter(body)
-	part, err := writer.CreateFormFile(c.param, file.Name())
-
-	r, err := ioutil.ReadAll(file)
-
-	_, err = part.Write(r)
-	if err != nil {
+	if _, err := io.Copy(part, r); err != nil {
 		return nil, err
 	}
-	err = writer.Close()
-	if err != nil {
+	if err := w.Close(); err != nil {
 		return nil, err
 	}
 
-	req, err := http.NewRequest("POST", c.endpoint, body)
+	req, err := http.NewRequest("POST", fmt.Sprintf("http://%v/convert", c.endpoint), buf)
```

Comment:
The service in https://github.com/sajari/docconv/blob/master/docd/main.go is currently HTTP only at the moment, can easily break out the pattern to be overridden though.